ros_packages/voice_assistant/voice_assistant/token_service.py

```diff
@@ -0,0 +1,168 @@
+import os
+import secrets
+from base64 import urlsafe_b64encode
+from hashlib import scrypt
+from typing import Union
+
+import rclpy
+from cryptography.fernet import Fernet
+from datatypes.srv import EncryptToken, DecryptToken, ExistToken
+from rclpy.executors import SingleThreadedExecutor
+from rclpy.node import Node
+from std_msgs.msg import String, Empty
+
+from . import SECRETS_DIR, SALT_PATH, TOKEN_PATH
+
+
+class TokenServiceNode(Node):
+    def __init__(self):
+        super().__init__("token_service")
+        self.active_token: Union[str, None] = None
+        self.is_token_stored: bool = self._check_if_previous_token_stored()
+
+        self.delete_token_subscription = self.create_subscription(
+            Empty, "delete_token", self.delete_token_callback, 10
+        )
+        self.token_exists_service = self.create_service(
+            ExistToken, "exist_token", self.token_exists_callback
+        )
+        self.encryption_service = self.create_service(
+            EncryptToken, "encrypt_token", self.encrypt_token_callback
+        )
+        self.decryption_service = self.create_service(
+            DecryptToken, "decrypt_token", self.decrypt_token_callback
+        )
+        self.token_publisher = self.create_publisher(String, "public_api_token", 10)
+
+        self.get_logger().info("Now running TOKEN SERVICE")
+
+    def publish_token(self, token: str) -> None:
+        self.active_token = token
+        msg = String()
+        msg.data = token
+        self.token_publisher.publish(msg)
+
+    def delete_token_callback(self, _: Empty) -> None:
+        try:
+            if self._check_if_previous_token_stored():
+                os.remove(TOKEN_PATH)
+                os.remove(SALT_PATH)
+        except Exception as e:
+            self.get_logger().warn(f"Token file could not be deleted: {e}")
+
+        # even if token stored in file cannot be deleted, remove the active token
+        self.is_token_stored = False
+        self.publish_token("")
+
+    def token_exists_callback(
+        self, _: ExistToken.Request, response: ExistToken.Response
+    ):
+        response.token_exists = self.is_token_stored
+        response.token_active = False
+        if self.active_token:
+            response.token_active = True
+        return response
+
+    def encrypt_token_callback(
+        self, request: EncryptToken.Request, response: EncryptToken.Response
+    ):
+        token: str = request.token
+        password: str = request.password
+
+        is_successful: bool = self.encrypt_token(token, password)
```

Comment:
For method calls I like to add the type hint for readability purposes, so I don't need to check what value the method returns. I also kept it for `token` and `password` because ROS datatypes are typically not inferred (at least by my IDE).